File: modules/vescape-core/ios/telemetry/AlertPresetGenerator.swift
import Foundation

struct PresetPoint: Decodable {
  let threshold: Double
  var thresholdMax: Double? = nil
  var repeatEverySeconds: Int64? = nil
}
struct PresetRelation {
  let fieldId: String
  let offset: Double
  let maxOffset: Double?
}
struct GeneratedPresetRule {
  let point: PresetPoint
  let soundType: String
  var relation: PresetRelation? = nil
}
struct PresetInput {
  var topSpeedKmh: Double = 50
  var hasBatteryConfig = false
  var speedUnitSystem = "metric"
  var matchBoardConfig = false
  var configBase: Double? = nil
}

private struct PresetDefinition: Decodable {
  enum Family: String, Decodable { case geiger, discrete }
  struct Range: Decodable {
    let start: Double
    let ceiling: Double
  }
  let soundType: String?
  let fieldId: String?
  let scaledByTopSpeed: Bool
  let requiresBatteryConfig: Bool
  let levels: [String: [PresetPoint]]

  enum CodingKeys: String, CodingKey { case family, soundType, fieldId, scaledByTopSpeed, requiresBatteryConfig, levels }
  init(from decoder: Decoder) throws {
    let values = try decoder.container(keyedBy: CodingKeys.self)
    soundType = try values.decodeIfPresent(String.self, forKey: .soundType)
    fieldId = try values.decodeIfPresent(String.self, forKey: .fieldId)
    scaledByTopSpeed = try values.decodeIfPresent(Bool.self, forKey: .scaledByTopSpeed) ?? false
    requiresBatteryConfig = try values.decodeIfPresent(Bool.self, forKey: .requiresBatteryConfig) ?? false
    let family = try values.decode(Family.self, forKey: .family)
    guard !scaledByTopSpeed || family == .geiger else { throw AlertPresetGenerator.InvalidPreset.invalidDefinition }
    switch family {
    case .geiger:
      let ranges = try values.decode([String: Range].self, forKey: .levels)
      levels = ranges.mapValues { [PresetPoint(threshold: $0.start, thresholdMax: $0.ceiling)] }
    case .discrete:
      levels = try values.decode([String: [PresetPoint]].self, forKey: .levels)
    }
    for level in AlertPresetGenerator.activeLevels {
      guard let points = levels[level], !points.isEmpty else { throw AlertPresetGenerator.InvalidPreset.invalidDefinition }
      for point in points {
        guard point.threshold.isFinite, point.thresholdMax?.isFinite != false,
              point.repeatEverySeconds.map({ $0 > 0 }) != false else { throw AlertPresetGenerator.InvalidPreset.invalidDefinition }
      }
    }
  }
}

/// Typed catalog and pure calculation shared by persisted presets and unsaved wizard previews.
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/telemetry/AlertPresetGenerator.kt
final class AlertPresetGenerator {
  static let metrics = ["speed", "duty", "motor-temp", "controller-temp", "battery"]
  static let activeLevels = ["safe", "normal", "minimal"]
  static let levels = activeLevels + ["off", "custom"]
  private static let kmhPerMph = 1.609344
  enum InvalidPreset: Error { case invalidIntent, invalidDefinition, missingResource }
  private struct Catalog: Decodable {
    let metrics: [String: PresetDefinition]
    let match: [String: PresetDefinition]
  }
  private struct CellCatalog: Decodable {
    struct Cell: Decodable { let id: String }
    let cells: [Cell]
  }
  private static func resource<T: Decodable>(_ name: String, as type: T.Type) throws -> T {
    #if SWIFT_PACKAGE
    let bundle = Bundle.module
    #else
    let module = Bundle(for: AlertPresetGenerator.self)
    let bundle = module.url(forResource: "VescapeCoreAssets", withExtension: "bundle").flatMap(Bundle.init(url:)) ?? module
    #endif
    guard let url = bundle.url(forResource: name, withExtension: "json") else { throw InvalidPreset.missingResource }
    return try JSONDecoder().decode(type, from: Data(contentsOf: url))
  }
  private static let catalog = Result { () throws -> Catalog in
    let value = try resource("alert-preset-definitions", as: Catalog.self)
    guard metrics.allSatisfy({ value.metrics[$0]?.soundType != nil }),
          value.match.allSatisfy({ metrics.contains($0.key) && $0.value.fieldId != nil }) else { throw InvalidPreset.invalidDefinition }
    return value
  }
  private static let cellIds = Result { Set(try resource("cell-presets", as: CellCatalog.self).cells.map(\.id)) }

  static func matchField(_ metric: String) throws -> String? { try catalog.get().match[metric]?.fieldId }

  static func validBattery(_ raw: Any?) throws -> Bool {
    guard let config = raw as? [String: Any] else { return false }
    switch config["mode"] as? String {
    case "manual":
      guard let minimum = config["minVoltage"] as? Double, let maximum = config["maxVoltage"] as? Double else { return false }
      return maximum > minimum
    case "preset":
      guard let id = config["cellPresetId"] as? String else { return false }
      return try cellIds.get().contains(id)
    default: return false
    }
  }

  static func generate(metric: String, level: String, input: PresetInput) throws -> [GeneratedPresetRule] {
    guard metrics.contains(metric), levels.contains(level), ["metric", "imperial"].contains(input.speedUnitSystem) else { throw InvalidPreset.invalidIntent }
    guard activeLevels.contains(level) else { return [] }
    let catalog = try catalog.get()
    guard let definition = catalog.metrics[metric] else { throw InvalidPreset.invalidDefinition }
    if input.matchBoardConfig, let match = catalog.match[metric] {
      return relativeRules(definition, match: match, level: level, base: input.configBase)
    }
    if definition.requiresBatteryConfig && !input.hasBatteryConfig { return [] }
    return fixedRules(definition, level: level, input: input)
  }

  private static func fixedRules(_ definition: PresetDefinition, level: String, input: PresetInput) -> [GeneratedPresetRule] {
    definition.levels[level]!.map { point in
      let threshold = definition.scaledByTopSpeed ? speedRange(point, input: input) : point
      return GeneratedPresetRule(point: threshold, soundType: definition.soundType!)
    }
  }

  private static func relativeRules(
    _ definition: PresetDefinition,
    match: PresetDefinition,
    level: String,
    base: Double?
  ) -> [GeneratedPresetRule] {
    match.levels[level]!.map { offset in
      let relation = PresetRelation(fieldId: match.fieldId!, offset: offset.threshold, maxOffset: offset.thresholdMax)
      let point: PresetPoint
      if let base {
        point = PresetPoint(
          threshold: roundTenth(base + offset.threshold),
          thresholdMax: offset.thresholdMax.map { roundTenth(base + $0) },
          repeatEverySeconds: offset.repeatEverySeconds
        )
      } else {
        // No anchor yet: persist the relationship, but no firing threshold.
        point = PresetPoint(threshold: 0, repeatEverySeconds: offset.repeatEverySeconds)
      }
      return GeneratedPresetRule(point: point, soundType: definition.soundType!, relation: relation)
    }
  }

  private static func speedRange(_ point: PresetPoint, input: PresetInput) -> PresetPoint {
    let top = input.topSpeedKmh.isFinite && input.topSpeedKmh > 0 ? input.topSpeedKmh : 50
    let start = roundTenth(point.threshold * top)
    let ceiling = roundTenth(point.thresholdMax! * top)
    if input.speedUnitSystem == "metric" { return PresetPoint(threshold: start, thresholdMax: ceiling) }
    let endMph = max(1, min(floor(top / kmhPerMph), floor(ceiling / kmhPerMph + 0.5)))
    let startMph = max(0, min(endMph - 1, floor(start / kmhPerMph + 0.5)))
    return PresetPoint(threshold: startMph * kmhPerMph, thresholdMax: endMph * kmhPerMph)
  }

  private static func roundTenth(_ value: Double) -> Double { floor(value * 10 + 0.5) / 10 }
}

File: modules/vescape-core/android/src/main/java/expo/modules/vescapecore/telemetry/AlertPresetGenerator.kt
package expo.modules.vescapecore.telemetry

import org.json.JSONObject
import kotlin.math.floor

internal data class PresetPoint(val threshold: Double, val thresholdMax: Double? = null, val repeatSeconds: Long? = null)
internal data class PresetRelation(val fieldId: String, val offset: Double, val maxOffset: Double?)
internal data class GeneratedPresetRule(val point: PresetPoint, val soundType: String, val relation: PresetRelation? = null)
internal data class PresetInput(
  val topSpeedKmh: Double = 50.0,
  val hasBatteryConfig: Boolean = false,
  val speedUnitSystem: String = "metric",
  val matchBoardConfig: Boolean = false,
  val configBase: Double? = null,
)

internal data class PresetDefinition(
  val soundType: String,
  val levels: Map<String, List<PresetPoint>>,
  val scaledByTopSpeed: Boolean,
  val requiresBatteryConfig: Boolean,
)
internal data class PresetMatchDefinition(val fieldId: String, val levels: Map<String, List<PresetPoint>>)

/** Typed catalog and pure calculation shared by saved presets and unsaved wizard previews.
 * @parity /modules/vescape-core/ios/telemetry/AlertPresetGenerator.swift
 */
internal object AlertPresetGenerator {
  val metrics = listOf("speed", "duty", "motor-temp", "controller-temp", "battery")
  val activeLevels = setOf("safe", "normal", "minimal")
  val levels = activeLevels + setOf("off", "custom")
  private const val KMH_PER_MPH = 1.609344

  private fun resource(name: String): JSONObject {
    val stream = checkNotNull(javaClass.getResourceAsStream("/$name.json")) { "Missing preset resource: $name" }
    return stream.bufferedReader().use { JSONObject(it.readText()) }
  }

  private val definitions: Map<String, PresetDefinition> by lazy {
    val catalog = resource("alert-preset-definitions").getJSONObject("metrics")
    metrics.associateWith { metric ->
      val value = catalog.getJSONObject(metric)
      val scaled = optionalBoolean(value, "scaledByTopSpeed")
      require(!scaled || value.getString("family") == "geiger") { "Scaled preset must be a range" }
      PresetDefinition(
        soundType = value.getString("soundType"),
        levels = parseLevels(value),
        scaledByTopSpeed = scaled,
        requiresBatteryConfig = optionalBoolean(value, "requiresBatteryConfig"),
      )
    }
  }
  private val matches: Map<String, PresetMatchDefinition> by lazy {
    val catalog = resource("alert-preset-definitions").getJSONObject("match")
    catalog.keys().asSequence().associateWith { metric ->
      require(metric in metrics) { "Unknown preset match metric: $metric" }
      val value = catalog.getJSONObject(metric)
      PresetMatchDefinition(value.getString("fieldId"), parseLevels(value))
    }
  }
  private val cellPresetIds: Set<String> by lazy {
    val cells = resource("cell-presets").getJSONArray("cells")
    (0 until cells.length()).map { cells.getJSONObject(it).getString("id") }.toSet()
  }

  private fun parseLevels(value: JSONObject): Map<String, List<PresetPoint>> {
    val levels = value.getJSONObject("levels")
    return activeLevels.associateWith { level ->
      when (value.getString("family")) {
        "geiger" -> {
          val range = levels.getJSONObject(level)
          listOf(PresetPoint(finiteNumber(range, "start"), finiteNumber(range, "ceiling")))
        }
        "discrete" -> {
          val points = levels.getJSONArray(level)
          require(points.length() > 0) { "Empty preset ladder" }
          (0 until points.length()).map { index ->
            val point = points.getJSONObject(index)
            val repeat = if (point.isNull("repeatEverySeconds")) null else finiteNumber(point, "repeatEverySeconds")
            require(repeat == null || (repeat > 0 && repeat == floor(repeat))) { "Invalid preset repeat interval" }
            PresetPoint(finiteNumber(point, "threshold"), repeatSeconds = repeat?.toLong())
          }
        }
        else -> error("Unknown preset family")
      }
    }
  }

  private fun optionalBoolean(value: JSONObject, key: String): Boolean {
    if (!value.has(key)) return false
    return value.get(key) as? Boolean ?: error("Invalid preset boolean: $key")
  }

  private fun finiteNumber(value: JSONObject, key: String): Double {
    val number = (value.get(key) as? Number)?.toDouble()
    require(number != null && number.isFinite()) { "Invalid preset number: $key" }
    return number
  }

  fun matchField(metric: String): String? = matches[metric]?.fieldId

  fun validBattery(raw: Any?): Boolean {
    val config = raw as? Map<*, *> ?: return false
    return when (config["mode"]) {
      "manual" -> {
        val min = (config["minVoltage"] as? Number)?.toDouble() ?: return false
        val max = (config["maxVoltage"] as? Number)?.toDouble() ?: return false
        max > min
      }
      "preset" -> config["cellPresetId"] in cellPresetIds
      else -> false
    }
  }

  fun generate(metric: String, level: String, input: PresetInput): List<GeneratedPresetRule> {
    require(metric in metrics) { "Unknown alert preset metric" }
    require(level in levels) { "Invalid preset level" }
    require(input.speedUnitSystem in setOf("metric", "imperial")) { "Invalid unit system" }
    if (level !in activeLevels) return emptyList()
    val definition = definitions.getValue(metric)
    val match = matches[metric].takeIf { input.matchBoardConfig }
    if (match != null) return relativeRules(definition, match, level, input.configBase)
    if (definition.requiresBatteryConfig && !input.hasBatteryConfig) return emptyList()
    return fixedRules(definition, level, input)
  }

  private fun fixedRules(
    definition: PresetDefinition,
    level: String,
    input: PresetInput,
  ): List<GeneratedPresetRule> =
    definition.levels.getValue(level).map { point ->
      val threshold = if (definition.scaledByTopSpeed) speedRange(point, input) else point
      GeneratedPresetRule(threshold, definition.soundType)
    }

  private fun relativeRules(
    definition: PresetDefinition,
    match: PresetMatchDefinition,
    level: String,
    base: Double?,
  ): List<GeneratedPresetRule> =
    match.levels.getValue(level).map { offset ->
      val relation = PresetRelation(match.fieldId, offset.threshold, offset.thresholdMax)
      val point = if (base == null) {
        // No anchor yet: persist the relationship, but no firing threshold.
        PresetPoint(threshold = 0.0, repeatSeconds = offset.repeatSeconds)
      } else {
        PresetPoint(
          threshold = roundTenth(base + offset.threshold),
          thresholdMax = offset.thresholdMax?.let { roundTenth(base + it) },
          repeatSeconds = offset.repeatSeconds,
        )
      }
      GeneratedPresetRule(point, definition.soundType, relation)
    }

  private fun speedRange(point: PresetPoint, input: PresetInput): PresetPoint {
    val top = input.topSpeedKmh.takeIf { it.isFinite() && it > 0 } ?: 50.0
    val start = roundTenth(point.threshold * top)
    val ceiling = roundTenth(checkNotNull(point.thresholdMax) * top)
    if (input.speedUnitSystem == "metric") return PresetPoint(start, ceiling)
    val endMph = maxOf(1.0, minOf(floor(top / KMH_PER_MPH), floor(ceiling / KMH_PER_MPH + 0.5)))
    val startMph = maxOf(0.0, minOf(endMph - 1, floor(start / KMH_PER_MPH + 0.5)))
    return PresetPoint(startMph * KMH_PER_MPH, endMph * KMH_PER_MPH)
  }

  private fun roundTenth(value: Double) = floor(value * 10 + 0.5) / 10
}
